### ISSUE TITLE
Loss function computation

### ISSUE THREAD

**Issue by rajeev595** [ruotianluo/self-critical.pytorch#110]
Please tell where the loss function is computed and also where the [_sample](https://github.com/ruotianluo/self-critical.pytorch/blob/9b36d9bf1cad47cd8c917c895d926b3ff3d3e5e8/models/FCModel.py#L150) function is called.

**Comment by ruotianluo** [comment]
Loss is in the loss_wrapper. Sample is called in forward.

**Comment by rajeev595** [comment]
Thanks for your prompt response. In the forward function, separate multinomial sampling is included but https://github.com/ruotianluo/self-critical.pytorch/blob/9b36d9bf1cad47cd8c917c895d926b3ff3d3e5e8/models/FCModel.py#L150 has both greedy and multinomial sampling which is not called anywhere. Please correct me if I am wrong.

**Comment by rajeev595** [comment]
While computing [loss](https://github.com/ruotianluo/self-critical.pytorch/blob/9b36d9bf1cad47cd8c917c895d926b3ff3d3e5e8/misc/loss_wrapper.py#L30), the inputs are `log(p_sampling)` , `samples`, `greedy_rewards`. So, [it](https://github.com/ruotianluo/self-critical.pytorch/blob/9b36d9bf1cad47cd8c917c895d926b3ff3d3e5e8/misc/utils.py#L95) is effectively computing the following: `- greedy_rewards * log(p_sampling)`

According to the paper, shouldn't it be `(sampling_rewards - greedy_rewards) * -log(p_sampling)` ?


Please help me understand.

**Comment by rajeev595** [comment]
While computing loss below https://github.com/ruotianluo/self-critical.pytorch/blob/9b36d9bf1cad47cd8c917c895d926b3ff3d3e5e8/misc/loss_wrapper.py#L30, the inputs are `log(p_sampling) , samples, greedy_rewards`. So, it is effectively computing `greedy_rewards * - log(p_sampling)` in the following:
https://github.com/ruotianluo/self-critical.pytorch/blob/9b36d9bf1cad47cd8c917c895d926b3ff3d3e5e8/misc/utils.py#L95

According to the paper, shouldn't it be `(sampling_rewards - greedy_rewards) * -log(p_sampling)` ?

Please help me understand.
